src/net.cpp

```diff
@@ -903,12 +918,31 @@ bool CConnman::AttemptToEvictConnection()
     // Protect 4 nodes that most recently sent us transactions.
     // An attacker cannot manipulate this metric without performing useful work.
     EraseLastKElements(vEvictionCandidates, CompareNodeTXTime, 4);
+    // Protect up to 8 non-tx-relay peers that have sent us blocks.
+    std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), CompareNodeBlockRelayOnlyTime);
+    size_t erase_size = std::min(size_t(8), vEvictionCandidates.size());
+    std::remove_if(vEvictionCandidates.end() - erase_size, vEvictionCandidates.end(), [](NodeEvictionCandidate const &n) { return !n.fRelayTxes && n.nLastBlockTime > 0; });
```

Comment:
@gmaxwell I should get rid of this `n.nLastBlockTime > 0` right?  So that we'll always protect up to 8 non-block-relay peers, even if they haven't managed to be first to send us a block.  But perhaps I should only protect them if `fRelevantServices` is true, so that non-full nodes do not get protected here?

Reply:
I think requring fRelevantServices makes sense.  I think it's not import to require that they sent a block-- helping them stay connected so long as they (claim they) could might help keep them around long enough to successfully do so.

Reply:
Fixed, thanks.